src/rotation.jl

```diff
@@ -0,0 +1,250 @@
+    end
+    kernel!(out, arr, sinθ, cosθ, midpoint, size(arr, 1), size(arr, 2),
+            ndrange=(size(arr, 1), size(arr, 2), size(arr, 3), size(arr, 4)))
+    return out
+end
+
+
+@kernel function imrotate_kernel_nearest!(out, arr, sinθ, cosθ, midpoint, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    _, _, _, _, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, midpoint, round) 
+    if 1 ≤ yrot_int ≤ imax && 1 ≤ xrot_int ≤ jmax
+        @inbounds out[i, j, c, b] = arr[yrot_int, xrot_int, c, b]
+    end
+end
+
+
+@kernel function imrotate_kernel_bilinear!(out, arr, sinθ, cosθ, midpoint, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+    
+    yrot, xrot, yrot_f, xrot_f, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, midpoint, floor) 
+    if 1 ≤ yrot_int ≤ imax - 1 && 1 ≤ xrot_int ≤ jmax - 1 
+
+        ydiff, ydiff_1minus, xdiff, xdiff_1minus = 
+            bilinear_helper(yrot, xrot, yrot_f, xrot_f)
+        @inbounds out[i, j, c, b] = 
+            (   xdiff_1minus    * ydiff_1minus  * arr[yrot_int      , xrot_int      , c, b]
+             +  xdiff_1minus    * ydiff         * arr[yrot_int + 1  , xrot_int      , c, b]
+             +  xdiff           * ydiff_1minus  * arr[yrot_int      , xrot_int + 1  , c, b] 
+             +  xdiff           * ydiff         * arr[yrot_int + 1  , xrot_int + 1  , c, b])
+    end
+end
+
+
+@kernel function ∇imrotate_kernel_nearest!(out, arr, sinθ, cosθ, midpoint, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    _, _, _, _, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, midpoint, round) 
+    if 1 ≤ yrot_int ≤ imax && 1 ≤ xrot_int ≤ jmax 
+        Atomix.@atomic out[yrot_int, xrot_int, c, b] += arr[i, j, c, b]
+    end
+end
+
+
+@kernel function ∇imrotate_kernel_bilinear!(out, arr, sinθ, cosθ, midpoint, imax, jmax)
+    i, j, c, b = @index(Global, NTuple)
+
+    yrot, xrot, yrot_f, xrot_f, yrot_int, xrot_int = rotate_coordinates(sinθ, cosθ, i, j, midpoint, floor) 
+    if 1 ≤ yrot_int ≤ imax - 1 && 1 ≤ xrot_int ≤ jmax - 1
+        o = arr[i, j, c, b]
+        ydiff, ydiff_1minus, xdiff, xdiff_1minus = 
+            bilinear_helper(yrot, xrot, yrot_f, xrot_f)
+        Atomix.@atomic out[yrot_int     ,   xrot_int    , c, b]  += xdiff_1minus    * ydiff_1minus * o
+        Atomix.@atomic out[yrot_int + 1 ,   xrot_int    , c, b]  += xdiff_1minus    * ydiff      * o
+        Atomix.@atomic out[yrot_int     ,   xrot_int + 1, c, b]  += xdiff           * ydiff_1minus * o
+        Atomix.@atomic out[yrot_int + 1 ,   xrot_int + 1, c, b]  += xdiff           * ydiff      * o
+    end
+end
+
+
+# is this rrule good? 
+# no @thunk and @unthunk
+function ChainRulesCore.rrule(::typeof(imrotate), array::AbstractArray{T}, θ; 
+                              method=:bilinear, midpoint=size(array) .÷ 2 .+ 1) where T
+    res = imrotate(array, θ; method, midpoint)
+    function pb_rotate(dy)
+        ad = ∇imrotate(unthunk(collect(dy)), θ; method, midpoint)
```

Comment:
I don't see why the desired array would be a CuArray or a normal Array. It should be possible to pass a `FillArray` to a KA kernel as long as its eltype `isbits` and the kernel doesn't try to mutate it. The problem here seems to be that `pb_rotate` assumes `dy` is not a fancy array type. A fix would be to pass the input `array` to `∇imrotate` as well. Then `_prepare_imrotate` can call similar on `array` to get an output of the correct type instead of relying on `similar(dy)` being the correct type.
```suggestion
    function pb_rotate(dy)
        ad = ∇imrotate(unthunk(dy), arr, θ; method, midpoint)
```